docs/proposals/20220210-md-node-label-sync.md

```diff
@@ -0,0 +1,161 @@
+---
+
+title: Label Sync Between MachineDeployment and underlying Kubernetes Nodes
+authors:
+  - "@arvinderpal"
+  - "@enxebre"
+reviewers:
+  - "@fabriziopandini"
+  - "@neolit123"
+  - "@sbueringer"
+  - "@vincepri"
+  - "@randomvariable"
+creation-date: 2022-02-10
+last-updated: 2022-02-26
+status: implementable
+replaces:
+superseded-by:
+
+---
+
+# Label Sync Between MachineDeployment and underlying Kubernetes Nodes
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+## Summary
+
+Labels on worker nodes are commonly used for node role assignment and workload placement. This document discusses a means by which labels placed on MachineDeployments can be kept in sync with labels on the underlying Kubernetes Nodes. 
+
+## Motivation
+
+Kubelet supports self-labeling of nodes via the `--node-labels` flag. CAPI users can specify these labels via `kubeletExtraArgs.node-labels` in the KubeadmConfigTemplate. There are a few shortcomings: 
+- Security concerns have restricted the allowed set of labels; In particular, kubelet's ability to self label in the `*.kubernetes.io/` and `*.k8s.io/` namespaces is mostly prohibited. Using prefixes outside these namespaces (i.e. unrestricted prefixes) is discouraged.  
+- Labels are only applicable at creation time. Adding/Removing a label would require a MachineDeployment rollout. This is undesirable especially in baremetal environments where provisioning can be time consuming. 
+
+The documented approach for specifying restricted labels in CAPI is to utilize kubectl. This introduces the potential for human error and also goes against the general declarative model adopted by CAPI.
+
+Users could also implement their own label synchronizer in their tooling, but this may not be practical for most. 
+
+### Goals
+
+- Support synchronizing labels that fall under the following restricted prefixes: `node-role.kubernetes.io/*` and `node-restriction.kubernetes.io/*`
+- Support synchronizing labels that fall under a CAPI specific prefix: `node.cluster.k8s.io/*`
+
+### Non-Goals/Future Work
+
+- Label sync for control-plane nodes. 
+- Support for arbitrary/user-specified label prefixes. 
+
+## Proposal
+
+CAPI will define its own label prefix which falls within the [restricted namespace *.k8s.io/](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/279-limit-node-access):
+
+```
+node.cluster.k8s.io/*
```

Comment:
Most of the CAPI labels are under `x-k8s.io`. Should we use this here as well?

Reply:
I think we're mixing up a few things. There are security concerns about the labels that kubelet sets. Specifically kubelet itself is only allowed to set a few labels because otherwise malicious kubelets could self-label and thus do bad things.

But we're discussion which labels we are setting as CAPI and I think in this case the following applies:
> The kubernetes.io/ and k8s.io/ prefixes are [reserved](https://kubernetes.io/docs/reference/labels-annotations-taints/) for Kubernetes core components.
> https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set

As we are not a Kubernetes core component we have to use our own namespace (like we did for all our other labels)


P.S. The comment order is interesting :) (yours 4 hours ago is above mine with 11 hours ago)

Reply:
I'll change the above to:
```
node.cluster.x-k8s.io/*
```

Reply:
Sounds good!